### PR TITLE
Simple event configuration

### PR DESCRIPTION
**Pull request by peterbryntesson**
feat: Added property TypeDefinitionId for the events - enabling us to do simple event configuration.

If the TypeDefinitionId is set then SelectClauses and WhereClause properties are ignored. Instead it sets up a select clause with all the fields in the type defined and base types. It also sets up a where clause that is OfType(TypeDefinitionId).

### CODE REVIEW

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs [hunk 1]
```diff
@@ -1181,12 +1183,690 @@ private class MonitoredItemWrapper {
...
+                return eventFilter;
+            }
+
             private HashSet<uint> _newTriggers = new HashSet<uint>();
             private HashSet<uint> _triggers = new HashSet<uint>();
             private Publisher.Models.MonitoringMode? _modeChange;
             private readonly ILogger _logger;
         }
 
+        internal static class ClientUtils {
```
Comment (simonjaeger) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs:1210]: Can we put this in a separate file?

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs [hunk 2]
```diff
@@ -1181,12 +1183,690 @@ private class MonitoredItemWrapper {
                 return fieldName;
             }
 
+            internal EventFilter BuildBasicEventFilter(Session session) {
+                var simpleEventTypeNodeId = NodeId.Parse(EventTemplate.EventFilter.EventTypeDefinitionId);
+                var fields = ClientUtils.CollectInstanceDeclarationsForType(session, simpleEventTypeNodeId);
+
+                var eventFilter = new EventFilter();
+                foreach (var field in fields) {
+                    eventFilter.SelectClauses.Add(new SimpleAttributeOperand() {
```
Comment (simonjaeger) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs:1192]: There is a AddSelectClause method on the EventFilter class. Should we use that?

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs [hunk 3]
```diff
@@ -1181,12 +1183,690 @@ private class MonitoredItemWrapper {
                 return fieldName;
             }
 
+            internal EventFilter BuildBasicEventFilter(Session session) {
```
Comment (simonjaeger) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs:1186]: GetDefaultEventFilter?

**Review by simonjaeger — CHANGES_REQUESTED** [review]

File: api/src/Microsoft.Azure.IIoT.Api/src/Publisher/Models/PublishedDataSetEventApiModel.cs [hunk 4]
```diff
@@ -80,5 +80,11 @@ public class PublishedDataSetEventApiModel {
         /// </summary>
         [DataMember(Name = "pendingAlarms", Order = 9)]
         public PendingAlarmsOptionsApiModel PendingAlarms { get; set; }
+
+        /// <summary>
+        /// Simple event Type id
+        /// </summary>
+        [DataMember(Name = "eventTypeDefinitionId", Order = 10)]
+        public string EventTypeDefinitionId { get; set; }
```
[the remark this answers is not in the document]
Comment (peterbryntesson) [reply api/src/Microsoft.Azure.IIoT.Api/src/Publisher/Models/PublishedDataSetEventApiModel.cs:88]: Maybe, let me think about it.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Extensions/PublishedDataSetEventModelEx.cs [hunk 5]
```diff
@@ -23,7 +23,11 @@ public static class PublishedDataSetEventModelEx {
         public static EventMonitoredItemModel ToMonitoredItem(
             this PublishedDataSetEventModel publishedEvent,
             string displayName = null) {
-            if (publishedEvent?.SelectClauses == null) {
+            if (publishedEvent == null) {
+                return null;
+            }
+
+            if (publishedEvent.SelectClauses == null && string.IsNullOrEmpty(publishedEvent.EventTypeDefinitionId)) {
```
[the remark this answers is not in the document]
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Extensions/PublishedDataSetEventModelEx.cs:30]: Isn't DisplayName part of SelectClause? So if you don't have any select clauses you don't have any DisplayNames

[the remark this answers is not in the document]
Comment (simonjaeger) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Extensions/PublishedDataSetEventModelEx.cs:30]: Yes that is one DisplayName. But there is also one for the entire event notifier that we should keep.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs (hunk 1 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs:1210]: We can, I just followed the pattern with many classes inside SubscriptionServices. But I can move it out.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs (hunk 2 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs:1192]: Good catch. I'll fix.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs (hunk 3 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs:1186]: Default is maybe not ideal either.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs (hunk 1 above)
Comment (simonjaeger) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs:1210]: Yeah, let's move it out. The file is already large.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs (hunk 3 above)
Comment (simonjaeger) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs:1186]: I agree. Tricky.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs (hunk 2 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs:1192]: AddSelectClause doesn't support BrowsePath as a collection. Need to be explicit.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Extensions/PublishedDataSetEventModelEx.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Extensions/PublishedDataSetEventModelEx.cs:30]: I'll try and remove the if check. Should be ok.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs (hunk 1 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs:1210]: Fixed.

File: api/src/Microsoft.Azure.IIoT.Api/src/Publisher/Models/PublishedDataSetEventApiModel.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (peterbryntesson) [reply api/src/Microsoft.Azure.IIoT.Api/src/Publisher/Models/PublishedDataSetEventApiModel.cs:88]: No, we need something else

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs (hunk 3 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs:1186]: GetSimpleEventFilter?

File: api/src/Microsoft.Azure.IIoT.Api/src/Publisher/Models/PublishedDataSetEventApiModel.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (peterbryntesson) [reply api/src/Microsoft.Azure.IIoT.Api/src/Publisher/Models/PublishedDataSetEventApiModel.cs:88]: Going to leave it with EventTypeDefinitionId for now.

[the remark this answers is not in the document]
Comment (simonjaeger) [reply api/src/Microsoft.Azure.IIoT.Api/src/Publisher/Models/PublishedDataSetEventApiModel.cs:88]: Can we remove Event from it? I think it's redundant as its already inside OpcEvents.

**Review by simonjaeger — APPROVED** [review]

File: api/src/Microsoft.Azure.IIoT.Api/src/Publisher/Models/PublishedDataSetEventApiModel.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (peterbryntesson) [reply api/src/Microsoft.Azure.IIoT.Api/src/Publisher/Models/PublishedDataSetEventApiModel.cs:88]: Yeah, maybe that is the right thing. Fixed.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Core/Extensions/EventFilterModelEx.cs [hunk 6]
```diff
@@ -49,6 +50,10 @@ public static class EventFilterModelEx {
             if (!model.WhereClause.IsSameAs(other.WhereClause)) {
                 return false;
             }
+            if (model.TypeDefinitionId != other.TypeDefinitionId) {
+                return false;
```
Comment (koepalex) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Core/Extensions/EventFilterModelEx.cs:54]: please remove tailing whitespaces

**Comment by koepalex** [comment]
> namespace Microsoft.Azure.IIoT.OpcUa.Protocol.Services {

please add missing copyright header

---
Refers to: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/ClientUtils.cs:1 in 76d5d6b. [](commit_id = 76d5d6b5f844720096e483ab73ec8b45db01da71, deletion_comment = False)

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Core/Extensions/EventFilterModelEx.cs (hunk 6 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Core/Extensions/EventFilterModelEx.cs:54]: Fixed

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/ClientUtils.cs [hunk 7]
```diff
@@ -0,0 +1,671 @@
...
+                        null,
+                        false,
+                        continuationPoints,
+                        out results,
+                        out diagnosticInfos);
+
+                    ClientBase.ValidateResponse(results, continuationPoints);
+                    ClientBase.ValidateDiagnosticInfos(diagnosticInfos, continuationPoints);
+                }
+                while (true);
```
Comment (koepalex) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/ClientUtils.cs:204]: We should have an timeout to not be forever in this loop

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/ClientUtils.cs [hunk 8]
```diff
@@ -0,0 +1,671 @@
+﻿// ------------------------------------------------------------
```
Comment (mregen) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/ClientUtils.cs:1]: Hi @peterbryntesson, for what are these utils needed?

Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/ClientUtils.cs:1]: I need to get all fields in the specified type and base types, so I can autopopulate a select clause.

Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/ClientUtils.cs:1]: I'm sure there is a better way to do this, I'm no expert on the OPC UA SDK. I guess you are the expert :-)

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/ClientUtils.cs (hunk 7 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/ClientUtils.cs]: Fixed

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/ClientUtils.cs (hunk 8 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/ClientUtils.cs]: @mregen updated the code to sue the NodeCache. Take a look at my latest commit.

**Review by mregen — APPROVED** [review]
:shipit:

**Pull request merged**